evm/src/cpu/kernel/asm/curve_add.asm

```diff
@@ -0,0 +1,417 @@
+// #define N 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47 // BN254 base field order
+
+global ec_add:
+    // Uncomment for test inputs.
+    // PUSH 0xdeadbeef
+    // PUSH 2
+    // PUSH 1
+    // PUSH 0x1bf9384aa3f0b3ad763aee81940cacdde1af71617c06f46e11510f14f3d5d121
+    // PUSH 0xe7313274bb29566ff0c8220eb9841de1d96c2923c6a4028f7dd3c6a14cee770
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+    DUP2
+    // stack: y0, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x0, y0, x0, y0, x1, y1, retdest
+    ISZERO
+    // stack: x0==0, y0, x0, y0, x1, y1, retdest
+    SWAP1
+    // stack: y0, x0==0, x0, y0, x1, y1, retdest
+    ISZERO
+    // stack: y0==0, x0==0, x0, y0, x1, y1, retdest
+    AND
+    // stack: y0==0 & x0==0, x0, y0, x1, y1, retdest
+    PUSH ec_add_first_zero
+    // stack: ec_add_first_zero, y0==0 & x0==0, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    DUP4
+    // stack: y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: x1, y1, x0, y0, x1, y1, retdest
+    ISZERO
+    // stack: x1==0, y1, x0, y0, x1, y1, retdest
+    SWAP1
+    // stack: y1, x1==0, x0, y0, x1, y1, retdest
+    ISZERO
+    // stack: y1==0, x1==0, x0, y0, x1, y1, retdest
+    AND
+    // stack: y1==0 & x1==0, x0, y0, x1, y1, retdest
+    PUSH ec_add_snd_zero
+    // stack: ec_add_snd_zero, y1==0 & x1==0, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    DUP4
+    // stack: y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: x1, y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: y0, x1, y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: x0, y0, x1, y1, x0, y0, x1, y1, retdest
+    %ec_check
+    // stack: isValid(x0, y0), x1, y1, x0, y0, x1, y1, retdest
+    PUSH ec_add_valid_first_point
+    // stack: ec_add_valid_first_point, isValid(x0, y0), x1, y1, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x1, y1, x0, y0, x1, y1, retdest
+    POP
+    // stack: y1, x0, y0, x1, y1, retdest
+    POP
+    // stack: x0, y0, x1, y1, retdest
+    POP
+    // stack: y0, x1, y1, retdest
+    POP
+    // stack: x1, y1, retdest
+    POP
+    // stack: y1, retdest
+    POP
+    // stack: retdest
```

Comment:
It looks like this returns nothing if the input is invalid? I think it would be hard for the caller (i.e. the wrapper we'll add which reads inputs from memory etc) to recognize error cases. We could return some special value such as `(u256::MAX, u256::MAX)`